### DISCUSSION TITLE
Non-secure context detected; cookies are not secured, and will not be available in cross-origin POST requests

### DISCUSSION THREAD

**Discussion by stere-marius** [keycloak/keycloak#36094] [Archive]
Hello,

I'm trying to resolve a warning related to Keycloak communicating with my backend over HTTP. Keycloak and my backend are on the same Docker network, and every time my backend communicates with Keycloak, we encounter the following warning:

`[org.keycloak.cookie.DefaultCookieProvider] (executor-thread-12) Non-secure context detected; cookies are not secured, and will not be available in cross-origin POST requests`

The issue stems from the fact that we are using HTTP to communicate with the Keycloak service within this Docker network. Specifically, I use http://keycloak:8080/auth to communicate with the Keycloak service.

Is there a way to resolve this issue without switching to HTTPS for internal communication between services?

Thank you!

**Comment by shawkins** [comment]
> Is there a way to resolve this issue without switching to HTTPS for internal communication between services?

This was also brought up on #35160. There isn't currently a straight-forward way to do this.

The current workaround without actually using HTTPS would be to use proxy headers such has x-forwarded-proto=https from your backend. Of course that means setting up the server proxy-headers variable and ensuring that only backend and actual proxies can actually communicate with Keycloak.

There is a related feature on realms called ssl required which is based upon checking if a local network was used to make a request. If something like that were configurable to your docker network and consulted for the secure context check, then it would work as you expect - but this may not be the direction we actually want to go. Any additional thoughs @vmuzikar @jonkoops @stianst ?

**Reply by jonkoops** [reply]
There is no way to fix this simply because this is enforced by the user-agent (browser). Keycloak attempts to set cookies in a way that will be most compatible, but we eventually are limited by what the user-agent allows.

**Comment by vmuzikar** [comment]
@stere-marius Could you please elaborate a bit on your specific use case? Especially why does your backend use cookies? Cookies are usually used only on the [frontend endpoints](https://www.keycloak.org/server/hostname#_background_server_endpoints) and there we could easily detect the actual protocol (whether it's HTTPS on the front even when the communication with the proxy is over HTTP – TLS edge termination) from the Hostname Provider, so we could easily enough not log the warning in these cases.

**Comment by MR4online** [comment]
Same issue here.
I try to access the keycloak REST API from an php docker container with the help of https://github.com/fschmtt/keycloak-rest-api-client-php (I'm happy for alternative suggestions, there doesn't seem to be many current packages).
Keycloak, NginX and PHP-App running in Container. SSL via Nginx. Access via Browser (UI) & Postman (API) works great. But not via PHP.
I tried a lot of different config-combinations, but none is working.
Can anyone help?

EDIT:
It's working with `baseUrl: 'keycloak:8080'` 
I realized the warning is realy just a warning and I failed using the PHP package. 
Anyway, this warning confused me, bacause I set HTTP and put cors to *

**Reply by shawkins** [reply]
@MR4online this discussion is for http access to Keycloak under two scenarios - either when there is an edge proxy and http access is coming not from the proxy, or possibly there's only [http access](https://github.com/keycloak/keycloak/discussions/36094#discussioncomment-11894717) 

If you have a cookie warning or seem to be failing cors checks when going through a proxy, then that means your keycloak proxy settings are incorrect. Please check your proxy-headers setting and ensure that the proxy is setting the appropriate headers.

> and put cors to *

Generally you don't want to do that.

**Comment by shawkins** [comment]
@stere-marius re-reading your discription are you in an http only deployment? That is the only access to keycloak is over http via the docker network?

**Comment by DanceMore** [comment]
I stumbled over similar errors:

```
2025-01-29 05:19:06,190 WARN  [org.keycloak.cookie.DefaultCookieProvider] (executor-thread-1) Non-secure context detected; cookies are not secured, and will not be available in cross-origin POST requests
2025-01-29 05:21:19,066 WARN  [org.keycloak.cookie.DefaultCookieProvider] (executor-thread-1) Non-secure context detected; cookies are not secured, and will not be available in cross-origin POST requests
```

I am attempting to deploy docker `image: keycloak/keycloak:26.0.0` in production mode without HTTPS because it is behind a Traefik Proxy that terminates SSL and performs LetsEncrypt for me.

these cookie warnings seem non-fatal so far, but I am performing upgrades from `image: jboss/keycloak:16.1.1`, so I am trying to make sure everything works in one environment before moving on to more. I do not intend to use any HTTP, even for service-to-service calls.

**Comment by DaBonse** [comment]
I get the same warning.

I use Keycloak behind a reverse proxy (local docker environment) with an self signed SSL cert. From the user view, everything works like expected. No Cookie warning is present.

In my application I also use several requests to REST Admin API in a non secure context. This means, I send a request from my app to Keycloak via http://host.docker.internal over port 80 direct to Keycloak without using the reverse proxy.
In this case I get the "Non-secure context detected" warning.

My REST API requests are plain PHP, over Guzzle which use curl in the end. The implementation didn't use cookies and the request didn't ask for it. For authentication my REST API client use the Bearer-Token.

Keycloak is configured in this way:

- KC_PROXY_HEADERS=xforwarded
- KC_HTTP_ENABLED=true


The non secure context is no problem for me, I secure the communication between the services by my own.

Of course, I can connect the API requests over the reverse proxy, this works and in this way there are no cookie warnings. But I would prefer a direct communication between the service containers.

**Reply by shawkins** [reply]
> The implementation didn't use cookies and the request didn't ask for it. For authentication my REST API client use the Bearer-Token.

The warning is emitted any time there is an interaction with the DefaultCookieProvider over a non-secure context. That can even happen just looking for cookies or trying to expire them, so for your scenario the warning may not actually be necessary. You can  log an enhancement request to limit when this warning is emitted to just when we are setting a non-secure cookie.

**Reply by dav1d-b** [reply]
> > The implementation didn't use cookies and the request didn't ask for it. For authentication my REST API client use the Bearer-Token.
> 
> The warning is emitted any time there is an interaction with the DefaultCookieProvider over a non-secure context. That can even happen just looking for cookies or trying to expire them, so for your scenario the warning may not actually be necessary. You can log an enhancement request to limit when this warning is emitted to just when we are setting a non-secure cookie.

It seems like the secure flag is always set to false unless the URI schema is set to HTTPS or when the host equals to localhost. It should probably also check X-Forwarded-Proto headers when evaluating the security context.

https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/utils/SecureContextResolver.java#L36-L70

**Reply by shawkins** [reply]
> It should probably also check X-Forwarded-Proto headers when evaluating the security context.

It does implicitly - if the proxy-headers option is configured and the headers are present, then the values reported by the request / connection will already take that into account.

**Reply by cobotrifork** [reply]
Hi @DaBonse . Have you figured anything about how to get less of the warnings without disabling the logs from the `DefaultCookieProvider` class?
If I understand it correctly from @shawkins, Keycloak does not support this use case completely?

**Reply by shawkins** [reply]
@cobotrifork there are two situations in which people are seeing this warning:
1. Misconfiguration behind a proxy - in an edge scenario you need the proxy-headers configured - https://github.com/keycloak/keycloak/issues/38091 will hopefully make this clearer in 26.3.
2. Trying to use Keycloak just over http with a non-localhost host name. The log warning is applicable to this scenario as the browser cannot infer a secure context is being used.

**Reply by cobotrifork** [reply]
Thanks for the quick reply!

> 1. Misconfiguration behind a proxy - in an edge scenario you need the proxy-headers configured - [Add more validation for proxy-headers #38091](https://github.com/keycloak/keycloak/issues/38091) will hopefully make this clearer in 26.3.

We did configure keycloak as in the OP with `KC_PROXY_HEADERS=xforwarded`. This covered the case where we access keycloak through the console.

> 2. Trying to use Keycloak just over http with a non-localhost host name. The log warning is applicable to this scenario as the browser cannot infer a secure context is being used.

I guess my issue is somewhat related to this? We also use `KC_HTTP_ENABLED=true` but that's because we run Keycloak inside a k8s cluster with istio and the in-cluster traffic is HTTP. 
However, there is no browser involved in the requests that cause this. Looking at traces, the requests come from backend services. For example a run of the https://github.com/adorsys/keycloak-config-cli from inside the cluster triggers this warning. I can't say I understand all the pieces of this puzzle , so any pointers to understand/fix this issue are much appreciated:)

**Reply by shawkins** [reply]
> However, there is no browser involved in the requests that cause this. Looking at traces, the requests come from backend services. For example a run of the https://github.com/adorsys/keycloak-config-cli from inside the cluster triggers this warning. I can't say I understand all the pieces of this puzzle , so any pointers to understand/fix this issue are much appreciated:)

Can you try this on Keycloak nightly? The fix for #38091 also tweaked when the secure context warning is emitted - it will only happen if we actually try to set a secure cookie, not just when the provider is initialized.

**Reply by cobotrifork** [reply]
I will try, but might take some time. Thanks for the pointer, I will let you know my results.

**Reply by cobotrifork** [reply]
@shawkins just tried it out. The nightly build does not log the warnings! Thanks for this, I guess we need to prepare to upgrade to 26.2 and 26.3 when it comes out.
Have a great weekend!:)

**Reply by shawkins** [reply]
@cobotrifork thank you for the confirmation. That will definitely help limit who encounters this warning moving forward.

**Reply by sweh** [reply]
This is not yet released in 26.2.5, is it?

**Reply by shawkins** [reply]
#38091 is for 26.3

**Reply by raoganeshr** [reply]
@shawkins I believe in 26.2, the code logs a warning even if cookie is blank.

In my case, keycloak is running as a pod inside k8s and my other services call on the http endpoint for permission checks. The token generated can be user tokens or service tokens. This warning seems to be logged everytime a call is made and looking at the code, it seems it just checks if the context is insecure but does not check if a cookie came as part of the request. Can this be patched for 26.2.X to make sure it only logs it if a cookie was present?
![image](https://github.com/user-attachments/assets/92e55c35-74d9-48e1-a0db-2ac488b36c64)

In main, the logic seems to have changed to log this when the cookie is actually set, which should reduce the frequency. However I am not sure what warned is for since I dont see it being stored anywhere.
![image](https://github.com/user-attachments/assets/aa17e519-3585-4d28-9eb5-ee7466fee544)

**Reply by shawkins** [reply]
> Can this be patched for 26.2.X to make sure it only logs it if a cookie was present?

It would take opening a new issue and a targetted change for 26.2 as the rest of the pr is too much for a back port. Would you want to try to do this? I think on the main branch we should also correct that it needs to look for a specific header, not just "X-Forwarded"

> However I am not sure what warned is for since I dont see it being stored anywhere.

What do you mean? The cookie is set on the response above this code block.

**Reply by franciscosousabr** [reply]
Was this really released in 26.3? I’m currently using 26.3.4 and I’m still receiving the warning. My scenario is very similar to the ones already reported here. I have Keycloak running in an ECS cluster, where I make introspection requests from an application that is also running on ECS. In this case, there is no reverse proxy between them, and there is no browser either. It’s just a POST request to the introspection API. Is there any specific configuration I need to apply?

**Reply by shawkins** [reply]
@franciscosousabr If you are getting this warning with no proxy, then it means you are using keycloak directly over http with a non-localhost address. The logic does consider that to be an insecure context.

If you want to discuss your setup further, a new discussion would be best.

**Reply by franciscosousabr** [reply]
The point is that the HTTP traffic occurs only within the internal network. For external traffic, the connection is HTTPS up to the Load Balancer, which terminates TLS and redirects HTTP to Keycloak, forwarding the headers (`--proxy-headers=forward`). This is working without any issues. The warning only appears when I perform token introspection from my API running on the same internal network, which is why I use Keycloak's internal address over HTTP, not HTTPS. However, this request does not use cookies; it’s just a `client_credentials` authentication with `client_id` and `secret_id` (`API → API`) with no browser involved. Therefore, I understand that this warning should not be displayed.

**Reply by shawkins** [reply]
> However, this request does not use cookies; it’s just a client_credentials authentication with client_id and secret_id (API → API) with no browser involved. 

The server is trying to set a cookie, that is triggering the warning.

**Reply by franciscosousabr** [reply]
So, for this scenario, the only way for me to handle it is to actually have my Keycloak responding over HTTPS, having to deal with self-signed certificates (since the cost of a certificate from an official CA wouldn’t make sense)?

**Reply by shawkins** [reply]
Is the problem that you are seeing a warning, or is something not working for you?

**Reply by franciscosousabr** [reply]
At first, everything is indeed working fine; the concern with the warning is that it’s affecting some monitoring. Is there any way I can disable this warning? Or perhaps some configuration I can set so that Keycloak doesn’t try to inject a cookie in the response, and therefore doesn’t fall into this scenario?

**Reply by shawkins** [reply]
You do have control over the log level for a given context, but you should probably look first at the flow for this authentication and see why it's setting a cookie.

**Comment by Pheggas** [comment]
Hey. I am getting the same error. I have keycloak and Caddy (reverse proxy) in docker environment on VM-B. I set up Caddy to make internal domain (keycloak.internal.my-domain.com) and forward the requests to keycloak-keycloak-1 container on port 8080 (Caddy and Keycloak has common docker network "proxy"). And every time i log in, it would throw this error / warning. When i want to publish the service (like shown in the [answer here](https://caddy.community/t/forward-to-service-under-another-caddy-setup/30156)) under keycloak.my-domain.com, it wouldn't even let me log in. After i enter credentials, it would hang on the loading circle and write these errors a couple of times. Where could be the issue?

PS: the in-depth description of my setup is in the link where i mentioned the answer.

**Comment by Sebastian-Gil-wcq** [comment]
At least the warning should be suppressed, but IMHO different deployments may face different challenges. We are using ISTIO TLS termination, and so far I only see the issue with logs being flooded with that error message. That said, assuming that environment is unsecure without the possibility to tell (override) Keycloak that actually we are in secure environment seems to be not correct. From my POV this is a regression in current version 26.1.3 in comparison to version 25.0.6 (the current version we are working with). As a minimum  the code should respect x-forwarded-proto=https

**Reply by shawkins** [reply]
> As a minimum the code should respect x-forwarded-proto=https

As mentioned here https://github.com/keycloak/keycloak/discussions/36094#discussioncomment-12141626 it does - please check your proxy-headers setting.

**Comment by gsustek** [comment]
And what about scenario if everything is http, the warning can not be and should not be avoided then, obviously?

**Reply by shawkins** [reply]
It may help to clarify if you mean an edge proxy scenario or just using keycloak directly over http. Either way we aren't checking to see if only have http enabled as a pre-condition for this warning. If you however use a localhost host over http, then there will be no warning.

**Reply by gsustek** [reply]
sorry not to be precise enough.
i meant: "just using keycloak directly over http"
Thnx.
